---
title: "Using basics of casc package"
author: "James Hopkins"
date: "February 2019"
package: "casc"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{How to use casc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warnings = FALSE,
  messages = FALSE,
  comment = "#>"
)
```


# Overview

A method to evaluate the number of clusters found in single cell RNAseq data. 
The idea is to select a the number of clusters with an acceptable ROC curve for each cluster.

A logistic regression model is trained with cells in half the provided clustering and an ROC curve is calculated with the remaining test data. 

# Installation

Installation with `devtools` package via:

```{r, eval=FALSE}
devtools::install_github("jamez-eh/casc")
```

 
# Basics

First load required packages.

```{r, results='hide', message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(casc)
library(SC3)
library(foreach)
library(scater)
```


## Load Data

To illustrate the usecase of `casc` we use a simulated dataset, included with the package, with 1000 cells and 500 genes. There are 10 simulated clusters, each with 20 markers.

```{r}
data(sce_sim)
data(test_sce)
```


## Clustering Data

The data must first be clustered. Here we are using the [SC3](http://bioconductor.org/packages/release/bioc/html/SC3.html) package to find clusters with values of k from 2 to 15.

```{r, results='hide', message=FALSE, warning=FALSE}
library(SC3)
sce_sim <- sc3(sce_sim, ks = 2:15, svm_max = 15000, gene_filter = FALSE)
```




## Casc Useage
[SC3](http://bioconductor.org/packages/release/bioc/html/SC3.html) stores clusters for various selection of k in slots `sc3_k_clusters`.
The function `cascer` takes a `singleCellExperiment` object and a list of clusterings and returns a list of `casc` objects. 
Each `casc` object can be accessed via `casc_list$casc_clusters_x`, where x is the index of the corresponding clustering in the list provided to `cascer`.

`casc` objects have 4 slots `predicted_classes`: classes predicted for test data
                            `auc`: mean auc for each classes ROC 
                            `response`: probabilities of each predicted class for test data
                            `truths`: true, provided classes for test data


```{r}
registerDoSEQ()
test_casc_list <- casc::cascer(sce_sim, list(sce_sim$sc3_2_clusters, sce_sim$sc3_3_clusters, sce_sim$sc3_5_clusters, sce_sim$sc3_8_clusters, sce_sim$sc3_9_clusters, sce_sim$sc3_10_clusters, sce_sim$sc3_11_clusters, sce_sim$sc3_12_clusters, sce_sim$sc3_15_clusters), alpha =0.5)


```

## Plot creation

A scatter plot of auc values for the list of `casc` objects can be created with the function `aucPlot`. This AUC is the mean of the 1 vs all binary AUCs. Multiple ROC curves can be visualized on the same graph by calling `multROCPlot` on a single `casc` object. `aucPlot` is for visualizing different values of k for the same clustering method. `multROCPlot` may be used to select between any clusterings.

```{r}
casc::aucPlot(test_casc_list)
multROCPlot(test_casc_list$casc_clusters_1)
multROCPlot(test_casc_list$casc_clusters_2)
casc::multROCPlot(test_casc_list$casc_clusters_3)
casc::multROCPlot(test_casc_list$casc_clusters_4)
casc::multROCPlot(test_casc_list$casc_clusters_5)
casc::multROCPlot(test_casc_list$casc_clusters_6)
casc::multROCPlot(test_casc_list$casc_clusters_7)
casc::multROCPlot(test_casc_list$casc_clusters_8)
casc::multROCPlot(test_casc_list$casc_clusters_9)
```
## Visualizing clusters
# Technical

```{r}
sessionInfo()
```










